### PR TITLE
Remote Reporters

### LINKED ISSUE #34: Collect Test Results from parallel tests / different tasty programs
Once we have #40 we can launch other tasty apps and collect test results directly from a sub tasty. We also can submit them via named pipes and collect them as nested test cases I think.
But then we need a way to execute nested tasty cases in a more controlled way (I think SimpleExec as a abstraction no longer will work here, but I'm not 100% sure about that).

#40 and #72 play in chare here.

I can think of a custom protocol here, but until `TestCaseResults` are a thing this would be difficult to solve

### PR DESCRIPTION
**Pull request by biohazard999**
fixes #34 

- [x] Add a remote reporter
- [x] Launch a tasty test
- [x] Inject Environment variables to connect those 2
- [x] Report back into the hosting process

I know this is again huge PR but to summarize there are a couple of things that are going on:

1. Extract transports into own plugins
1. introduce a new `InMemoryTransport`
1. introduce the `TastyCommander` which controls a scope or a remote scope e.g. `TastyProcessCommander` and `TastyRemoteCommander`
1. Refactor and cleanup the tests to match more our style guidelines
1. Introduce a new overload to `TastyScope` to allow reporting nested test cases given an `IAsyncEnumerable<TestCaseResult>`
1. Fixed a few bugs on the go for registering some plugins
1. Removed the `SimpleExec` dependency from `Tasty.Core`
1. Fixed the integration tests to all report success

This will the last PR needed to start dogfooding `Tasty` in `Xenial`

### CODE REVIEW

File: src/Xenial.Tasty.Core/Tasty.Scope.cs [hunk 1]
```diff
@@ -322,9 +322,23 @@ public async Task<int> Run(string[] args)
                 await PluginLoader.LoadPlugins(this).ConfigureAwait(false);
             }
             var executor = new TestExecutor(this);
+
+            foreach (var executorMiddleWare in executorMiddleWares)
+            {
+                executorMiddleWare(executor);
```
Comment (Paramethod) [inline src/Xenial.Tasty.Core/Tasty.Scope.cs:328]: Casing: MiddleWares- > Middlewares

File: src/Xenial.Tasty.Transports.NamedPipes/Transports/NamedPipesTranport.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+using Xenial.Delicious.Remote;
+
+namespace Xenial.Delicious.Transports
+{
+    public static class NamedPipesTranport
+    {
+        public static Task<TransportStreamFactory?> CreateNamedPipeTransportStream(CancellationToken token = default)
+        {
+            var connectionType = Environment.GetEnvironmentVariable(EnvironmentVariables.InteractiveConnectionType);
+            if (string.Equals(connectionType, "NamedPipes", StringComparison.InvariantCultureIgnoreCase))
```
Comment (Paramethod) [inline src/Xenial.Tasty.Transports.NamedPipes/Transports/NamedPipesTranport.cs:18]: If the environment variable is not found or is doesn't equal to NamedPipes... should it throw? How does the user knows it failed?

Comment (biohazard999) [reply src/Xenial.Tasty.Transports.NamedPipes/Transports/NamedPipesTranport.cs:18]: This is a fall through protocol, so the next transport registered will look if a connection can be established. (For example Websockets).
Not every session is interactive (this is what I'm currently working on) so this needs to be handled by the `ConnectToRemoteStreamMiddleware` and throw there.
I will convert this to a `ConnectionString` property.

As for now the user never will get involved into building this connection string though, only during an internal abstraction.

File: src/Xenial.Tasty.Core/Tasty.Scope.cs [hunk 3]
```diff
@@ -72,6 +74,12 @@ public TastyScope RegisterCommand(Func<(string name, Func<RuntimeContext, Task>
             return RegisterCommand(name, command, description, isDefault ?? false);
         }
 
+        public TastyScope Use(Action<TestExecutor> executorMiddleWare)
```
Comment (Paramethod) [inline src/Xenial.Tasty.Core/Tasty.Scope.cs:77]: Casing: MiddleWare -> Middleware

File: src/Xenial.Tasty.Core/Remote/TastyRemoteDefaults.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    public delegate Task<ITastyRemote> ConnectToRemote(TastyScope scope, Stream remoteStream);
+
+    public static class TastyRemoteDefaults
+    {
+        public static Task<bool> IsInteractiveRun()
+        {
+            var isInteractive = Environment.GetEnvironmentVariable(EnvironmentVariables.InteractiveMode);
+            if (!string.IsNullOrEmpty(isInteractive))
+            {
+                if (bool.TryParse(isInteractive, out var result))
```
Comment (Paramethod) [inline src/Xenial.Tasty.Core/Remote/TastyRemoteDefaults.cs:28]: bool.TryParse is safe with null or empty. Will return false.

File: src/Xenial.Tasty.Core/Commander.Tasty.cs [hunk 5]
```diff
@@ -117,43 +137,69 @@ private async Task ReportSummary(IEnumerable<TestCaseResult> testCases)
...
+
+                jsonRpc = JsonRpc.Attach(stream, tastyServer);
+                disposables.Add(jsonRpc);
+                running = true;
+                return tastyServer;
+            }
+            return null;
+        }
+
+        public async IAsyncEnumerable<TestCaseResult> WaitForResults()
```
Comment (biohazard999) [inline src/Xenial.Tasty.Core/Commander.Tasty.cs:164]: Currently there is no way for timeouts. We should address this in another issue.

File: src/Xenial.Tasty.Cli/Views/MainWindow.ViewModel.cs [hunk 6]
```diff
@@ -24,24 +26,32 @@ internal class MainWindowViewModel : IDisposable
...
-            Commander = new TastyCommander()
-                .UseNamedPipesTransport()
-                .RegisterReporter(Report)
-                .RegisterReporter(ReportSummary);
+            var connectionString = NamedPipesConnectionStringBuilder.CreateNewConnection();
 
-            LogProgress = new Progress<(string line, bool isRunning, int exitCode)>(p =>
+            LogProgress = new Progress<(string line, bool isError, int? exitCode)>(p =>
             {
                 LogText += $"{p.line.TrimEnd(Environment.NewLine.ToArray())}{Environment.NewLine}";
```
Comment (Paramethod) [inline src/Xenial.Tasty.Cli/Views/MainWindow.ViewModel.cs:43]: StringBuilder?

Comment (biohazard999) [reply src/Xenial.Tasty.Cli/Views/MainWindow.ViewModel.cs]: So the StringBuilder will go away anyway

**Pull request merged**